File: 503. Next Greater Element II.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
var nextGreaterElements = function(nums) {
    let [stk, ans] = [[], new Array(nums.length).fill(-1)];

    for(let i=0;i<2;i++){
        for(const [cur, cur_num] of Object.entries(nums)){
            while(stk.length>0 && cur_num>stk[stk.length-1][1]){
                let [past, past_num] = stk.pop();
                ans[past] = cur_num;
            }
            stk.push([cur, cur_num]);
        }
    }
    return ans;
};

File: 503. Next Greater Element II.py
class Solution:
    def nextGreaterElements(self, nums: List[int]) -> List[int]:
        stk, ans = [], [-1]*len(nums)
        
        for i in range(2):
            for cur, cur_num in enumerate(nums):
                while stk and cur_num>stk[-1][1]:
                    past, past_num = stk.pop()
                    ans[past] = cur_num
            
                stk.append((cur, cur_num))
        
        return ans